### PR TITLE
RFC for concepts redesign.

### PR DESCRIPTION
**Pull request by dom96**
This has been a long time coming. Refs #13.

### CODE REVIEW

File: RFCs/Concepts_Redesign.md [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+For the other examples suggested by @zah I would like to see more concrete
+use cases. I wouldn't be surprised if they can be implemented with the syntax
+I have in mind as well but I don't understand them enough to be able to
+provide equivalent examples.
+
+## Backward incompatibility
+
+This is a backwards incompatible change but concepts are an experimental feature
+which does not hold stability guarantees (even as of the upcoming v1.0).
```
Comment (Clyybber) [inline RFCs/Concepts_Redesign.md]: According to @Araq on IRC they do. (except for `do`)

File: RFCs/Concepts_Redesign.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+understanding a concept far easier to developers. The biggest problem with the
+current concepts is that they need to be evaluated, evaluating a concept when
+trying to determine why it doesn't match your type can be really time consuming
+and frustrating. What this RFC proposes will make it significantly easier to
+understand what a concept specifies and to figure out why a type doesn't
+satisfy its spec.
+
+The downsides to this proposal are that what can be specified is diminished,
+that said I don't believe this is significant and I go through examples of this
+below.
```
Comment (Clyybber) [inline RFCs/Concepts_Redesign.md]: This RFC should not remove the "old" concepts, IMO.

File: RFCs/Concepts_Redesign.md (hunk 1 above)
Comment (dom96) [reply RFCs/Concepts_Redesign.md]: Link? Everything that's in experimental has no stability guarantees, we wrote this in the release notes.

File: RFCs/Concepts_Redesign.md (hunk 2 above)
Comment (dom96) [reply RFCs/Concepts_Redesign.md]: Not immediately, but eventually it should.

File: RFCs/Concepts_Redesign.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+This RFC proposes a simpler concept syntax, one that cuts right to the point of
+what a concept specifies.
+
+
+## Motivation
+
+The current concepts syntax is difficult to grok as it requires mentally
+evaluating Nim code. The concepts implementation is also currently very buggy
+and would require significant effort to make reliable.
```
Comment (Clyybber) [inline RFCs/Concepts_Redesign.md]: Probably a similar effort required to make the concepts this RFC proposes reliable. In fact most if not all cases these new concepts cover are reliable with the current concepts.

File: RFCs/Concepts_Redesign.md (hunk 1 above)
Comment (Clyybber) [reply RFCs/Concepts_Redesign.md]: https://irclogs.nim-lang.org/06-09-2019.html#21:34:38

Comment (Clyybber) [reply RFCs/Concepts_Redesign.md]: Not that I necessarily agree that all experimental features should have stability guarantees..

Comment (Araq) [reply RFCs/Concepts_Redesign.md]: Er, to clarify, what I meant was these features won't be removed, but overhauled. 'do' is an exception, we can simply remove it, without loss of functionality.

File: RFCs/Concepts_Redesign.md [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+   for f in fields(x):
+      f.serialize(Stream)
+```
+
+After:
+
+```nim
+type
+  Serializable[T] = concept
+    iterator fields(x: Serializable): T
```
Comment (Araq) [inline RFCs/Concepts_Redesign.md]: No, that's wrong. It's not a single ``T`` for an object with fields of different types.

Comment (dom96) [reply RFCs/Concepts_Redesign.md]: oh oops, indeed.

File: RFCs/Concepts_Redesign.md (hunk 1 above)
Comment (dom96) [reply RFCs/Concepts_Redesign.md]: That's not true at all as far as our promise in the release notes goes. You shouldn't be making extra promises in such contexts, as far as the promise we made on nim-lang.org any experimental feature can be removed or replaced by another feature.

File: RFCs/Concepts_Redesign.md (hunk 4 above)
Comment (dom96) [reply RFCs/Concepts_Redesign.md]: I think a recursive definition will work, let me know what you think :)

**Comment by zah** [comment]
I feel like a broken record by now, but all my points from the previous discussion apply here as well:
https://github.com/nim-lang/RFCs/issues/13#issuecomment-298886380

Point 1 (that concepts must be able to require associated types and constants) is especially important. Any design that doesn't cover this aspect is just not workable.

In general, the type requirements in generic programming are often fuzzier than the typical OOP interface and I don't see a particularly good reason to limit the expressive power of the language, only to satisfy some superficial syntax preferences. If you think that the current implementation issues of concepts are caused by their syntax, you are deeply mistaken.

File: RFCs/Concepts_Redesign.md (hunk 4 above)
Comment (zah) [reply RFCs/Concepts_Redesign.md]: So, this is resolved? I'd like to see the explanation in the manual that describes the solution.

**Comment by zah** [comment]
Also, concepts are now finally standardised in C++.

Is anyone of the opponents of the current syntax familiar with the C++ design? Our current design is pretty close to C++'s way of doing things and soon millions of developers will be trained to understand this syntax. I wonder if anyone would have ever complained if you were first trained in your university to understand what a concept definition is (in C++).

**Comment by dom96** [comment]
@zah I have referred to your comment explicitly in my document and have read it. I did my best to show what a Serializable concept would look like with this syntax, for the other examples I need more information as I didn't really understand their purpose.

As far as the syntax goes, I think my argument around it is far from superficial, the fact that concepts currently require evaluating the concept body (both by the compiler and by a human in their head) is a big downside. But if that really doesn't create implementation problems then I think we should at least instead provide the syntax I propose on top of the current concepts via a macro or otherwise.

**Comment by Araq** [comment]
But this

```nim

type
  Serializable = concept
    iterator fields(x: Serializable): RootObj
    proc serialize(f: Serializable, s: Stream)
```

doesn't make sense.

**Comment by Araq** [comment]
Superseeded by my RFC.

**Pull request closed without merging**